File: rag-quickstart/typescript/src/config-loader.ts
/**
 * Load tessera_config.json by walking up to the repo root, or fall back to
 * documented defaults so the example runs standalone.
 *
 * The Tessera Launchpad writes tessera_config.json into the repository root.
 * Examples live in rag-quickstart/<lang>/, so we search upward from the start
 * directory to find it.
 */

import { existsSync, readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";

const CONFIG_FILENAME = "tessera_config.json";
const SUPPORTED_SCHEMA_VERSION = 1;

export interface TesseraConfig {
  host: string;
  port: number;
  apiKey: string;
  secure: boolean;
}

export interface EmbeddingConfig {
  provider: string;
  endpoint: string;
  apiKey: string;
  model: string;
  deploymentName: string;
  dimension: number;
}

export interface IngestionConfig {
  tenantId: string;
  chunkSize: number;
  chunkOverlap: number;
  dataDir: string;
}

export interface Config {
  tessera: TesseraConfig;
  embedding: EmbeddingConfig;
  ingestion: IngestionConfig;
}

function defaultConfig(): Config {
  return {
    tessera: { host: "localhost", port: 50051, apiKey: "", secure: false },
    embedding: {
      provider: "ollama",
      endpoint: "http://localhost:11434",
      apiKey: "",
      model: "nomic-embed-text",
      deploymentName: "",
      dimension: 768,
    },
    ingestion: {
      tenantId: "rag-quickstart",
      chunkSize: 800,
      chunkOverlap: 100,
      dataDir: "./data",
    },
  };
}

function findConfigFile(startDir: string): string | null {
  let current = resolve(startDir);
  for (;;) {
    const candidate = join(current, CONFIG_FILENAME);
    if (existsSync(candidate)) {
      return candidate;
    }
    const parent = dirname(current);
    if (parent === current) {
      return null;
    }
    current = parent;
  }
}

function parseConfig(data: any): Config {
  if (data.schema_version !== SUPPORTED_SCHEMA_VERSION) {
    throw new Error(
      `Unsupported schema_version ${JSON.stringify(data.schema_version)}; ` +
        `this example supports schema_version ${SUPPORTED_SCHEMA_VERSION}.`,
    );
  }
  const t = data.tessera;
  const e = data.embedding;
  const i = data.ingestion;
  return {
    tessera: {
      host: t.host,
      port: Number(t.port),
      apiKey: t.api_key ?? "",
      secure: Boolean(t.secure ?? false),
    },
    embedding: {
      provider: e.provider,
      endpoint: e.endpoint ?? "",
      apiKey: e.api_key ?? "",
      model: e.model ?? "",
      deploymentName: e.deployment_name ?? "",
      dimension: Number(e.dimension),
    },
    ingestion: {
      tenantId: i.tenant_id,
      chunkSize: Number(i.chunk_size),
      chunkOverlap: Number(i.chunk_overlap),
      dataDir: i.data_dir ?? "./data",
    },
  };
}

export function loadConfig(startDir: string = process.cwd()): Config {
  const configPath = findConfigFile(startDir);
  if (configPath === null) {
    return defaultConfig();
  }
  const data = JSON.parse(readFileSync(configPath, "utf-8"));
  return parseConfig(data);
}

File: rag-quickstart/python/config_loader.py
"""Load ermya_config.json by walking up to the repo root, or fall back to
documented defaults so the example runs standalone.

The Ermya Launchpad writes ermya_config.json into the repository root when
it generates this project. Examples live in rag-quickstart/<lang>/, so we search
upward from the start directory to find it.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

CONFIG_FILENAME = "ermya_config.json"
SUPPORTED_SCHEMA_VERSION = 1


@dataclass
class ErmyaConfig:
    host: str
    port: int
    api_key: str
    secure: bool


@dataclass
class EmbeddingConfig:
    provider: str
    endpoint: str
    api_key: str
    model: str
    deployment_name: str
    dimension: int


@dataclass
class IngestionConfig:
    tenant_id: str
    chunk_size: int
    chunk_overlap: int
    data_dir: str


@dataclass
class VlsUser:
    username: str
    password: str


@dataclass
class VlsConfig:
    issuer: str
    token_endpoint: str
    client_id: str
    users: dict[str, VlsUser]


@dataclass
class Config:
    ermya: ErmyaConfig
    embedding: EmbeddingConfig
    ingestion: IngestionConfig
    vls: VlsConfig | None = None


def _default_config() -> Config:
    """Documented defaults: local Ermya + local Ollama, no API keys required."""
    return Config(
        ermya=ErmyaConfig(
            host="localhost", port=50051, api_key="", secure=False
        ),
        embedding=EmbeddingConfig(
            provider="ollama",
            endpoint="http://localhost:11434",
            api_key="",
            model="nomic-embed-text",
            deployment_name="",
            dimension=768,
        ),
        ingestion=IngestionConfig(
            tenant_id="rag-quickstart",
            chunk_size=800,
            chunk_overlap=100,
            data_dir="./data",
        ),
    )


def _find_config_file(start_dir: Path) -> Path | None:
    current = start_dir.resolve()
    while True:
        candidate = current / CONFIG_FILENAME
        if candidate.is_file():
            return candidate
        if current.parent == current:
            return None
        current = current.parent


def _parse_config(data: dict) -> Config:
    version = data.get("schema_version")
    if version != SUPPORTED_SCHEMA_VERSION:
        raise ValueError(
            f"Unsupported schema_version {version!r}; "
            f"this example supports schema_version {SUPPORTED_SCHEMA_VERSION}."
        )
    t = data["ermya"]
    e = data["embedding"]
    i = data["ingestion"]
    vls = None
    vls_raw = data.get("vls")
    if vls_raw is not None:
        vls = VlsConfig(
            issuer=vls_raw["issuer"],
            token_endpoint=vls_raw["token_endpoint"],
            client_id=vls_raw["client_id"],
            users={
                owner: VlsUser(username=u["username"], password=u["password"])
                for owner, u in vls_raw["users"].items()
            },
        )
    return Config(
        ermya=ErmyaConfig(
            host=t["host"],
            port=int(t["port"]),
            api_key=t.get("api_key", ""),
            secure=bool(t.get("secure", False)),
        ),
        embedding=EmbeddingConfig(
            provider=e["provider"],
            endpoint=e.get("endpoint", ""),
            api_key=e.get("api_key", ""),
            model=e.get("model", ""),
            deployment_name=e.get("deployment_name", ""),
            dimension=int(e["dimension"]),
        ),
        ingestion=IngestionConfig(
            tenant_id=i["tenant_id"],
            chunk_size=int(i["chunk_size"]),
            chunk_overlap=int(i["chunk_overlap"]),
            data_dir=i.get("data_dir", "./data"),
        ),
        vls=vls,
    )


def load_config(start_dir: Path | None = None) -> Config:
    """Return the parsed config, or documented defaults if no file is found."""
    start = Path(start_dir) if start_dir is not None else Path.cwd()
    config_path = _find_config_file(start)
    if config_path is None:
        return _default_config()
    data = json.loads(config_path.read_text())
    return _parse_config(data)
